prow/cmd/exporter/collector.go

```diff
@@ -169,3 +171,153 @@ func sanitizeLabelName(s string) string {
 func escapeString(v string) string {
 	return escapeWithDoubleQuote.Replace(v)
 }
+
+type prowJobLifecycleCollector struct {
+	lastKnownState map[types.UID]lastState
+	HistogramVec   *prometheus.HistogramVec
+}
+
+type lastState struct {
+	state    prowapi.ProwJobState
+	observed time.Time
+}
+
+func (pjc *prowJobLifecycleCollector) Update(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) {
+	now := time.Now()
+	if lastKnownState, exists := pjc.lastKnownState[newJob.UID]; !exists {
+		if newJob.Status.State == prowapi.TriggeredState {
+			now = newJob.CreationTimestamp.Time
+			pjc.lastKnownState[newJob.UID] = lastState{
+				state:    newJob.Status.State,
+				observed: now,
+			}
+			return
+		} else if oldJob == nil || oldJob.Status.State == newJob.Status.State {
+			// we can't determine the time when the job entered the current state, do nothing
+			// TODO: we can reconstruct some states by looking at the pods
+			return
+		}
+		// first time we see a state transition for a job live, or the job is still pending, now we record the time
+		pjc.lastKnownState[newJob.UID] = lastState{
+			state:    newJob.Status.State,
+			observed: now,
+		}
+		return
+	} else if lastKnownState.state != newJob.Status.State {
+		pjc.lastKnownState[newJob.UID] = lastState{
+			state:    newJob.Status.State,
+			observed: now,
+		}
+
+		labels := getJobLabel(lastKnownState, newJob)
+		histogram, err := pjc.HistogramVec.GetMetricWithLabelValues(labels.values()...)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to get a histogram for a prowjob")
+			return
+		}
+
+		histogram.Observe(now.Sub(lastKnownState.observed).Seconds())
+	}
+}
+
+func (pjc *prowJobLifecycleCollector) Delete(job *prowapi.ProwJob) {
+	if _, exists := pjc.lastKnownState[job.UID]; exists {
+		delete(pjc.lastKnownState, job.UID)
+	}
+}
+
+// NewProwJobLifecycleCollector creates histograms which can track the timespan between ProwJob state transitions.
+// The histograms are based on the job name, the old job state and the new job state.
+// Data is collected by hooking itself into the prowjob informer.
+// The collector will never record the same state transition twice, even if reboots happen.
+// The collector will not record the first state transition it sees after booting, and therefore miss the first transition of each job after a restart.
+// TODO: by looking at the created pods we can collect more transitions on restarts
+func NewProwJobLifecycleCollector(informer cache.SharedIndexInformer) *prowJobLifecycleCollector {
```

Comment:
In order to keep this a bit more tidy, can we put the `ProwJobLifecycleCollector` into its own package?

Reply:
Yes